<a href="https://colab.research.google.com/github/Divasdev/-100-Days-of-Code-Python-Pro-Bootcamp/blob/master/DataAnalysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [13]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [14]:
df=pd.read_csv('/content/bestsellers.csv')
df

,Name,Author,User Rating,Reviews,Price,Year,Genre
0,10-Day Green Smoothie Cleanse,JJ Smith,4.7,17350,8,2016,Non Fiction
1,11/22/63: A Novel,Stephen King,4.6,2052,22,2011,Fiction
2,12 Rules for Life: An Antidote to Chaos,Jordan B. Peterson,4.7,18979,15,2018,Non Fiction
3,1984 (Signet Classics),George Orwell,4.7,21424,6,2017,Fiction
4,"5,000 Awesome Facts (About Everything!) (Natio...",National Geographic Kids,4.8,7665,12,2019,Non Fiction
...,...,...,...,...,...,...,...
545,Wrecking Ball (Diary of a Wimpy Kid Book 14),Jeff Kinney,4.9,9413,8,2019,Fiction
546,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2016,Non Fiction
547,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2017,Non Fiction
548,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2018,Non Fiction


In [15]:
df.head(5)


,Name,Author,User Rating,Reviews,Price,Year,Genre
0,10-Day Green Smoothie Cleanse,JJ Smith,4.7,17350,8,2016,Non Fiction
1,11/22/63: A Novel,Stephen King,4.6,2052,22,2011,Fiction
2,12 Rules for Life: An Antidote to Chaos,Jordan B. Peterson,4.7,18979,15,2018,Non Fiction
3,1984 (Signet Classics),George Orwell,4.7,21424,6,2017,Fiction
4,"5,000 Awesome Facts (About Everything!) (Natio...",National Geographic Kids,4.8,7665,12,2019,Non Fiction


In [16]:
df.tail(5)

,Name,Author,User Rating,Reviews,Price,Year,Genre
545,Wrecking Ball (Diary of a Wimpy Kid Book 14),Jeff Kinney,4.9,9413,8,2019,Fiction
546,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2016,Non Fiction
547,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2017,Non Fiction
548,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2018,Non Fiction
549,You Are a Badass: How to Stop Doubting Your Gr...,Jen Sincero,4.7,14331,8,2019,Non Fiction


In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Name         550 non-null    object 
 1   Author       550 non-null    object 
 2   User Rating  550 non-null    float64
 3   Reviews      550 non-null    int64  
 4   Price        550 non-null    int64  
 5   Year         550 non-null    int64  
 6   Genre        550 non-null    object 
dtypes: float64(1), int64(3), object(3)
memory usage: 30.2+ KB


In [18]:
df.describe()

,User Rating,Reviews,Price,Year
count,550.000000,550.000000,550.000000,550.000000
mean,4.618364,11953.281818,13.100000,2014.000000
std,0.226980,11731.132017,10.842262,3.165156
min,3.300000,37.000000,0.000000,2009.000000
25%,4.500000,4058.000000,7.000000,2011.000000
50%,4.700000,8580.000000,11.000000,2014.000000
75%,4.800000,17253.250000,16.000000,2017.000000
max,4.900000,87841.000000,105.000000,2019.000000


Prize Ranges From MIN-:0 to MAX-:150

Reviews Some books have Review as min as 37 to some having max review of 87841.It is Extremely wide range

user rating starting from 3.3 as the min to max rating of 4.9

PHASE 1

In [23]:
df['Genre'].unique()

array(['Non Fiction', 'Fiction'], dtype=object)

In [26]:
df['Genre'].nunique()

2

In [25]:
df['Genre'].value_counts()

,count
Genre,
Non Fiction,310
Fiction,240


In [28]:
df['Name'].unique()

array(['10-Day Green Smoothie Cleanse', '11/22/63: A Novel',
       '12 Rules for Life: An Antidote to Chaos',
       '1984 (Signet Classics)',
       '5,000 Awesome Facts (About Everything!) (National Geographic Kids)',
       'A Dance with Dragons (A Song of Ice and Fire)',
       'A Game of Thrones / A Clash of Kings / A Storm of Swords / A Feast of Crows / A Dance with Dragons',
       'A Gentleman in Moscow: A Novel',
       'A Higher Loyalty: Truth, Lies, and Leadership',
       'A Man Called Ove: A Novel',
       "A Patriot's History of the United States: From Columbus's Great Discovery to the War on Terror",
       'A Stolen Life: A Memoir', 'A Wrinkle in Time (Time Quintet)',
       'Act Like a Lady, Think Like a Man: What Men Really Think About Love, Relationships, Intimacy, and Commitment',
       'Adult Coloring Book Designs: Stress Relief Coloring Book: Garden Designs, Mandalas, Animals, and Paisley Patterns',
       'Adult Coloring Book: Stress Relieving Animal Designs',


In [29]:
df['Name'].nunique()

351

In [30]:
df['Name'].value_counts()

,count
Name,
"Publication Manual of the American Psychological Association, 6th Edition",10
StrengthsFinder 2.0,9
"Oh, the Places You'll Go!",8
The Very Hungry Caterpillar,7
The 7 Habits of Highly Effective People: Powerful Lessons in Personal Change,7
...,...
What to Expect When You're Expecting,1
What Should Danny Do? (The Power to Choose Series),1
What Pet Should I Get? (Classic Seuss),1


In [31]:
df.isnull()

,Name,Author,User Rating,Reviews,Price,Year,Genre
0,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...
545,False,False,False,False,False,False,False
546,False,False,False,False,False,False,False
547,False,False,False,False,False,False,False
548,False,False,False,False,False,False,False


In [34]:
if df.isnull():
  df['Price']=df['Price'].fillna(df['Price'].mean())
  df['Reviews']=df['Reviews'].fillna(df['Reviews'].mean())

In [35]:
df.duplicated()

,0
0,False
1,False
2,False
3,False
4,False
...,...
545,False
546,False
547,False
548,False


In [36]:
df[df['Name'] == 'Publication Manual of the American Psychological Association, 6th Edition']

,Name,Author,User Rating,Reviews,Price,Year,Genre
271,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2009,Non Fiction
272,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2010,Non Fiction
273,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2011,Non Fiction
274,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2012,Non Fiction
275,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2013,Non Fiction
276,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2014,Non Fiction
277,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2015,Non Fiction
278,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2016,Non Fiction
279,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2017,Non Fiction
280,Publication Manual of the American Psychologic...,American Psychological Association,4.5,8580,46,2018,Non Fiction


In [40]:
df.filter(regex='Name')


,Author
0,JJ Smith
1,Stephen King
2,Jordan B. Peterson
3,George Orwell
4,National Geographic Kids
...,...
545,Jeff Kinney
546,Jen Sincero
547,Jen Sincero
548,Jen Sincero


In [41]:
df.filter(regex='Author')

,Author
0,JJ Smith
1,Stephen King
2,Jordan B. Peterson
3,George Orwell
4,National Geographic Kids
...,...
545,Jeff Kinney
546,Jen Sincero
547,Jen Sincero
548,Jen Sincero


In [44]:
high_rated_books=df[df['User Rating']>4.8]
high_rated_books

,Name,Author,User Rating,Reviews,Price,Year,Genre
40,"Brown Bear, Brown Bear, What Do You See?",Bill Martin Jr.,4.9,14344,5,2017,Fiction
41,"Brown Bear, Brown Bear, What Do You See?",Bill Martin Jr.,4.9,14344,5,2019,Fiction
81,Dog Man and Cat Kid: From the Creator of Capta...,Dav Pilkey,4.9,5062,6,2018,Fiction
82,Dog Man: A Tale of Two Kitties: From the Creat...,Dav Pilkey,4.9,4786,8,2017,Fiction
83,Dog Man: Brawl of the Wild: From the Creator o...,Dav Pilkey,4.9,7235,4,2018,Fiction
84,Dog Man: Brawl of the Wild: From the Creator o...,Dav Pilkey,4.9,7235,4,2019,Fiction
85,Dog Man: Fetch-22: From the Creator of Captain...,Dav Pilkey,4.9,12619,8,2019,Fiction
86,Dog Man: For Whom the Ball Rolls: From the Cre...,Dav Pilkey,4.9,9089,8,2019,Fiction
87,Dog Man: Lord of the Fleas: From the Creator o...,Dav Pilkey,4.9,5470,6,2018,Fiction
146,"Goodnight, Goodnight Construction Site (Hardco...",Sherri Duskey Rinker,4.9,7038,7,2012,Fiction


In [51]:
filter_under_10=df[(df['Genre']=='Fiction') & (df['Price']<10)]
filter_under_10

,Name,Author,User Rating,Reviews,Price,Year,Genre
3,1984 (Signet Classics),George Orwell,4.7,21424,6,2017,Fiction
9,A Man Called Ove: A Novel,Fredrik Backman,4.6,23848,8,2016,Fiction
10,A Man Called Ove: A Novel,Fredrik Backman,4.6,23848,8,2017,Fiction
13,A Wrinkle in Time (Time Quintet),Madeleine L'Engle,4.5,5153,5,2018,Fiction
40,"Brown Bear, Brown Bear, What Do You See?",Bill Martin Jr.,4.9,14344,5,2017,Fiction
...,...,...,...,...,...,...,...
541,Wonder,R. J. Palacio,4.8,21625,9,2014,Fiction
542,Wonder,R. J. Palacio,4.8,21625,9,2015,Fiction
543,Wonder,R. J. Palacio,4.8,21625,9,2016,Fiction
544,Wonder,R. J. Palacio,4.8,21625,9,2017,Fiction


In [54]:
filter_under_10.value_counts().sum()

np.int64(129)

In [56]:
auth_name='Fredrik Backman'
specific_author=df[df['Author']==auth_name]
specific_author

,Name,Author,User Rating,Reviews,Price,Year,Genre
9,A Man Called Ove: A Novel,Fredrik Backman,4.6,23848,8,2016,Fiction
10,A Man Called Ove: A Novel,Fredrik Backman,4.6,23848,8,2017,Fiction


In [60]:
specific_book=df[(df['Year']==2019) & (df['Reviews']>10000)]
specific_book

,Name,Author,User Rating,Reviews,Price,Year,Genre


Let's inspect the `df` by checking each condition separately to understand why the combined filter resulted in an empty DataFrame.

In [61]:
# Books published in 2019
books_2019 = df[df['Year'] == 2019]
display(books_2019.head())

,Name,Author,User Rating,Reviews,Price,Year,Genre
4,"5,000 Awesome Facts (About Everything!) (Natio...",National Geographic Kids,4.8,7665,12,2019,Non Fiction
33,Becoming,Michelle Obama,4.8,61133,11,2019,Non Fiction
41,"Brown Bear, Brown Bear, What Do You See?",Bill Martin Jr.,4.9,14344,5,2019,Fiction
44,Can't Hurt Me: Master Your Mind and Defy the Odds,David Goggins,4.8,16244,18,2019,Non Fiction
74,Difficult Riddles For Smart Kids: 300 Difficul...,M Prefontaine,4.6,7955,5,2019,Non Fiction


In [62]:
# Books with more than 10,000 reviews
books_high_reviews = df[df['Reviews'] > 10000]
display(books_high_reviews.head())

,Name,Author,User Rating,Reviews,Price,Year,Genre
0,10-Day Green Smoothie Cleanse,JJ Smith,4.7,17350,8,2016,Non Fiction
2,12 Rules for Life: An Antidote to Chaos,Jordan B. Peterson,4.7,18979,15,2018,Non Fiction
3,1984 (Signet Classics),George Orwell,4.7,21424,6,2017,Fiction
5,A Dance with Dragons (A Song of Ice and Fire),George R. R. Martin,4.4,12643,11,2011,Fiction
6,A Game of Thrones / A Clash of Kings / A Storm...,George R. R. Martin,4.7,19735,30,2014,Fiction


By examining the outputs above, you can see if any specific books appear in both filtered lists. If a book exists in both, then the original `specific_book` filter *should* have returned it. If `specific_book` is still empty, it implies that the intersection of these two conditions is indeed empty for your dataset, or there might be an issue with data types that isn't immediately apparent from `df.info()`.